docs/content/guides/dagster/pythonic-resources.mdx

```diff
@@ -0,0 +1,440 @@
+class MyIOManager(ConfigurableIOManager):
+    root_path: str
+
+    def _get_path(self, asset_key: AssetKey) -> str:
+        return self.root_path + "/".join(asset_key.path)
+
+    def handle_output(self, context: OutputContext, obj):
+        write_csv(self._get_path(context.asset_key), obj)
+
+    def load_input(self, context: InputContext):
+        return read_csv(self._get_path(context.asset_key))
+
+defs = Definitions(
+    assets=...,
+    resources={
+        "io_manager": MyIOManager(root_path="/tmp/")
+    },
+)
+```
+
+---
+
+## Advanced resource patterns
+
+The following sections describe more advanced patterns for using Pythonic resources.
+
+### Adapting function-style resources
+
+In codebases which utilize [function-style](https://docs.dagster.io/concepts/resources#resources) resources with the <PyObject object="@resource"/> decorator, you can use the <PyObject object="ConfigurableResourceAdapter"/> class to adapt these resources to the new Pythonic resource pattern.
+
+The adapter subclass should list all config fields as attributes, and should implement a `wrapped_resource` property which returns the underlying resource definition to adapt.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_resource_adapter endbefore=end_resource_adapter
+# Old code, cannot be changed for back-compat purposes
+class Writer:
+    def __init__(self, prefix: str):
+        self._prefix = prefix
+
+    def output(self, text: str) -> None:
+        print(self._prefix + text)
+
+
+@resource(config_schema={"prefix": str})
+def writer_resource(context):
+    prefix = context.resource_config["prefix"]
+    return Writer(prefix)
+
+
+# New adapter layer
+class WriterResource(ConfigurableResourceAdapter):
+    prefix: str
+
+    @property
+    def wrapped_resource(self) -> ResourceDefinition:
+        return writer_resource
+
+
+@asset
+def my_asset(writer: Writer):
+    writer.output("hello, world!")
+
+
+defs = Definitions(assets=[my_asset], resources={"writer": WriterResource(prefix="greeting: ")})
+```
+
+### Using bare Python objects as resources
+
+In some cases, you may want to use a bare Python object as a resource which is not a subclass of <PyObject object="ConfigurableResource"/>.
+
+For example, you may want to directly pass a third-party API client into your assets or ops. This follows a similar pattern to using a <PyObject object="ConfigurableResource"/> subclass, however assets and ops which use these resources must [annotate](https://docs.python.org/3/library/typing.html#typing.Annotated) them with <PyObject object="Resource"/>.
+
+```python file=/guides/dagster/pythonic_resources/pythonic_resources.py startafter=start_raw_github_resource endbefore=end_raw_github_resource
+@asset
+def public_github_repos(github: Resource[GitHub]):
+    return github.organization("dagster-io").repositories()
```

Comment:
Recommend adding comment here to emphasize how annotations work
e.g.

`# Resource[Github] is treated exactly like Github for type checking purposes and the runtime type of the github parameter is GitHub. The purpose of Resource is to tell dagster that github is a resource rather than an upstream asset.`